addons/mass_mailing/models/mail_notification.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from dateutil.relativedelta import relativedelta
+
+from odoo import api, fields, models
+
+
+class MailNotification(models.Model):
+    # Extend the mail.notification model for tracking (marketing purpose)
+    _inherit = 'mail.notification'
+
+    # mail data
+    mail_id_int = fields.Integer(
+        string='Mail ID (tech)',
+        help='ID of the related mail_mail. This field is an integer field because '
+             'the related mail_mail can be deleted separately from its statistics. '
+             'However the ID is needed for several action and controllers.',
+        index=True)
+    email = fields.Char(string="Email")
+    message_id = fields.Char(string='Message-ID', help="Technical field for the email Message-ID (RFC 2392)")
+    # document
+    model = fields.Char(string='Document model')
+    res_id = fields.Many2oneReference(string='Document ID', model_field='model')
+    # campaign / wave data
+    # if mass_mailing_id is set the notification will act as a trace
+    mass_mailing_id = fields.Many2one('mailing.mailing', string='Mailing', index=True, ondelete='cascade')
+    campaign_id = fields.Many2one('utm.campaign', related='mass_mailing_id.campaign_id', string='Campaign', index=True, store=True)
+    # Bounce and tracking
+    canceled = fields.Datetime(help='Date when the email has been invalidated. '
+                                    'Invalid emails are blacklisted, opted-out or invalid email format')
+    scheduled = fields.Datetime(help='Date when the email has been created', default=fields.Datetime.now)
+    sent = fields.Datetime(help='Date when the email has been sent')
+    exception = fields.Datetime(help='Date of technical error leading to the email not being sent')
+    opened = fields.Datetime(help='Date when the email has been opened the first time')
+    replied = fields.Datetime(help='Date when this email has been replied for the first time.')
+    bounced = fields.Datetime(help='Date when this email has bounced.')
+    # Link tracking
+    links_click_ids = fields.One2many('link.tracker.click', 'mailing_trace_id', string='Links click')
+    clicked = fields.Datetime(help='Date when customer clicked on at least one tracked link')
+    # Status
+    notification_status = fields.Selection(selection_add=[
+        ('outgoing', 'Outgoing'),
+        ('opened', 'Opened'),
+        ('replied', 'Replied')])
+    status_update = fields.Datetime('State Update', help='Last state update of the mail')
+
+    def name_get(self):
+        return [(n.id, '%s: %s (%s)' % (n.notification_type, n.mass_mailing_id.name or n.res_partner_id.display_name or '', n.id)) for n in self]
```

Comment:
Meh, I usually don't like one letter variables, even for that kind of structure.

Suggestion:

```
return [(notification.id, '%s: %s (%s)' % (
    notification.notification_type,
    notification.mass_mailing_id.name or notification.res_partner_id.display_name or '',
    notification.id
)) for notification in self]
```